### PR TITLE
Replace the ConfigureCookieOptions action property with the CookieBuilder

### PR DESCRIPTION
**Pull request by natemcmaster**
Addresses https://github.com/aspnet/HttpAbstractions/issues/853.

Uses the new CookieBuilder API added here: https://github.com/aspnet/HttpAbstractions/pull/882

cc @JunTaoLuo @rynowak @javiercn

### CODE REVIEW

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs [hunk 1]
```diff
@@ -14,81 +14,78 @@ public class AntiforgeryOptions
...
+        [Obsolete("This property is obsolete and will be removed in a future version. The recommended alternative is " + nameof(Cookie) + "." + nameof(CookieBuilder.Domain) + ".")]
+        public string CookieDomain { get => Cookie.Domain; set => Cookie.Domain = value; }
 
         /// <summary>
-        /// Configures the <see cref="CookieOptions"/> of the antiforgery cookies. Without additional configuration, the 
-        /// default values antiforgery cookie options are true for <see cref="CookieOptions.HttpOnly"/>, null for
-        /// <see cref="CookieOptions.Domain"/> and <see cref="SameSiteMode.Strict"/> for <see cref="CookieOptions.SameSite"/>.
+        /// Determines the settings used to create the antiforgery cookies.
         /// </summary>
-        public Action<HttpContext, CookieOptions> ConfigureCookieOptions { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: Does cookie builder have something similar to this?

**Comment by rynowak** [comment]
:shipit: this looks great, I just want to double confirm that the builder has something like our configure delgate that's getting removed.

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs (hunk 1 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: No, but we made `CookieBuilder.Build(HttpContext context)` virtual and `AntiforgeryOptions.Cookie` settable for edge cases where users need to do something with HttpContext that we don't expect.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: We added this for a reason, so we'll have to figure out an alternative.

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: What was the reason? When we were designing CookieBuilder, we figured anything other than the already-available top-level properties is an uncommon usage of this API.

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs [hunk 2]
```diff
@@ -111,5 +108,20 @@ public string FormFieldName
...
+            public AntiforgeryCookieBuilder()
+            {
+                SameSite = SameSiteMode.Strict;
+                HttpOnly = true;
+            }
+
+            public override string Name
+            {
+                get => base.Name;
+                set => base.Name = value ?? throw new ArgumentNullException(nameof(value));
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: null or empty?

Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: 1.x behavior only checks for null and only throws ArgumentNullException.

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs [hunk 3]
```diff
@@ -69,22 +69,14 @@ public void SaveCookieToken(HttpContext httpContext, string token)
...
-#pragma warning restore 618
-                SameSite = SameSiteMode.Strict,
-                Secure = _options.RequireSsl
-            };
-
-#pragma warning disable 618
-            if (_options.CookiePath != null)
+            var options = _options.Cookie.Build(httpContext);
+
+            options.Secure = options.Secure || _options.RequireSsl;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: Why not depend on CookieSecurePolicy?

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs [hunk 4]
```diff
@@ -14,81 +14,78 @@ public class AntiforgeryOptions
...
-        public Action<HttpContext, CookieOptions> ConfigureCookieOptions { get; set; }
+        /// <remarks>
+        /// <para>
+        /// If an explicit <see cref="CookieBuilder.Name"/> is not provided, the system will automatically generate a
+        /// unique name that begins with <see cref="DefaultCookiePrefix"/>.
+        /// </para>
+        /// <para>
+        /// <see cref="CookieBuilder.SameSite"/> defaults to <see cref="SameSiteMode.Strict"/>.
+        /// <see cref="CookieBuilder.HttpOnly"/> defaults to <c>true</c>.
+        /// <see cref="CookieBuilder.SecurePolicy"/> will be forced to <seealso cref="CookieSecurePolicy.Always"/> when <see cref="RequireSsl"/> is <c>true</c>.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: Map RequireSsl to Cookie.SecurePolicy?

File: src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/AntiforgeryOptions.cs]: I know we did this to enable per-request configuration of things like the path, but I can't find an actual customer request for this, so I think we can just forget it until someone asks. Like you said, build is virtual, so we could use our subclass to resurrect this if necessary

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: This has historically been here along with `RequireSsl`. I'm not really sure if the reasons are important or not.

**Comment by rynowak** [comment]
:shipit: from me

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs (hunk 3 above)
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgeryTokenStore.cs]: I looked through the code, and didn't find any significant reason for these to be separate options. I've pushed an update that obsoletes `RequireSsl` and forwards it to `Cookie.SecurePolicy`.

**Comment by rynowak** [comment]
RequireSsl does more than just configure the cookie policy https://github.com/aspnet/Antiforgery/blob/f258be61fd05ff23d286661c58409b1eba004440/src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs#L268

File: src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs [hunk 5]
```diff
@@ -265,12 +265,11 @@ private void SaveCookieTokenAndHeader(HttpContext httpContext, string cookieToke
 
         private void CheckSSLConfig(HttpContext context)
         {
-            if (_options.RequireSsl && !context.Request.IsHttps)
+            if (_options.Cookie.SecurePolicy == CookieSecurePolicy.Always && !context.Request.IsHttps)
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs]: cc @rynowak

Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs]: RequireSsl does more than just configure the cookie policy https://github.com/aspnet/Antiforgery/blob/f258be61fd05ff23d286661c58409b1eba004440/src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs#L268

Comment (rynowak) [inline src/Microsoft.AspNetCore.Antiforgery/Internal/DefaultAntiforgery.cs]: Ok cool 👍

**Pull request merged**